### ISSUE TITLE
cmd/6l: Go 1.4.2 version of 6l causes malloc assertion failure

### ISSUE THREAD

**Issue by bha123** [golang/go#12738]
Hi all

We are installing golang 1.4.2 on ubuntu14.04 . After using the instructions https://golang.org/doc/install?download=go1.4.2.linux-amd64.tar.gz

when i am running simple helloworld program . It is giving the below error

```
root@ip-10-164-196-28:/home/vcap/goworkspace# go test 1.go
# command-line-arguments
6l: malloc.c:3096: sYSMALLOc: Assertion `(old_top == (((mbinptr) (((char *) &((av)->bins[((1) - 1) * 2])) - __builtin_offsetof (struct malloc_chunk, fd)))) && old_size == 0) || ((unsigned long) (old_size) >= (unsigned long)((((__builtin_offsetof (struct malloc_chunk, fd_nextsize))+((2 * (sizeof(size_t))) - 1)) & ~((2 * (sizeof(size_t))) - 1))) && ((old_top)->size & 0x1) && ((unsigned long)old_end & pagemask) == 0)' failed.
```

regards
Bharath

**Comment by davecheney** [comment]
@bha123 can you please share a code sample that others can use to reproduce this issue.

**Comment by ianlancetaylor** [comment]
Please tell us the results of

type go
go env
ls $GOROOT

**Comment by ianlancetaylor** [comment]
Also

go test -x 1.go

**Comment by ianlancetaylor** [comment]
I just verified that the Go installation works for me on Ubuntu 14.04.

**Comment by bha123** [comment]
Hi iamlancetaylor

The following are the results i got
vcap@ip-10-32-58-8:~$ type go
go is hashed (/usr/local/go/bin/go)
go env

vcap@ip-10-32-58-8:~$ sudo go env
GOARCH="amd64"
GOBIN=""
GOCHAR="6"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH=""
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="g++"
CGO_ENABLED="1"

vcap@ip-10-32-58-8:~$ sudo go test -x 1.go
WORK=/tmp/go-build928749057
mkdir -p $WORK/command-line-arguments/_obj/
mkdir -p $WORK/command-line-arguments/_obj/exe/
cd /home/vcap
/usr/local/go/pkg/tool/linux_amd64/6g -o $WORK/command-line-arguments.a -trimpath $WORK -p command-line-arguments -complete -D _/home/vcap -I $WORK -pack ./1.go
cd .
/usr/local/go/pkg/tool/linux_amd64/6l -o $WORK/command-line-arguments/_obj/exe/a.out -L $WORK -extld=gcc $WORK/command-line-arguments.a
# command-line-arguments
6l: malloc.c:3096: sYSMALLOc: Assertion `(old_top == (((mbinptr) (((char *) &((av)->bins[((1) - 1) * 2])) - __builtin_offsetof (struct malloc_chunk, fd)))) && old_size == 0) || ((unsigned long) (old_size) >= (unsigned long)((((__builtin_offsetof (struct malloc_chunk, fd_nextsize))+((2 * (sizeof(size_t))) - 1)) & ~((2 * (sizeof(size_t))) - 1))) && ((old_top)->size & 0x1) && ((unsigned long)old_end & pagemask) == 0)' failed.


regards
Bharath

**Comment by ianlancetaylor** [comment]
Why are you running "sudo go"?  What happens if you omit the "sudo"?

What is the output of "printenv"?

What is the contents of the file 1.go?

**Comment by rsc** [comment]
@bha123, can you please answer Ian's questions? 

Also if you run `go test -work ...` then the go command will leave the $WORK directory behind (will not remove it). After that, can you please run:

	$ WORK=<line from output>
	$ gdb --args /usr/local/go/pkg/tool/linux_amd64/6l -o $WORK/command-line-arguments/_obj/exe/a.out -L $WORK -extld=gcc $WORK/command-line-arguments.a
	(gdb) r
	... program crashes in malloc, hopefully ...
	(gdb) where

It may be interesting to see what the stack trace is at the failure. Another possibility would be to run `valgrind` instead of `gdb --args` and see if valgrind detects any misuse of memory from within 6l.

Thanks.

**Comment by rsc** [comment]
(Normally we would deprioritize bugs in old Go releases, but if there is an actual memory corruption bug in 6l we need to fix it to keep Go 1.4 being a valid bootstrap toolchain.)

**Comment by rsc** [comment]
Like Ian, I cannot reproduce this on Ubuntu 14.04. I even changed the go command to invoke 'valgrind -q 6l' instead of '6l' and was able to run all.bash successfully. Given the lack of further information, I don't think there's more we can do here.
